src/proto/single/cmd.rs

```diff
@@ -245,6 +244,31 @@ impl FaktoryCommand for Push {
 
 // ----------------------------------------------
 
+pub struct PushBulk<'a>(&'a [Job]);
```

Comment:
Feels weird for this to hold references to `Job` when `Push` holds an owned `Job`. See the suggestion on `enqueue_many` for one alternative.

In the next major version we may want to make both this and `Push` refer to `&Job` (or maybe `Cow<'_, Job>`?) so that callers don't _have_ to pass in owned `Job` given that `issue` doesn't require it.

Reply:
As of now we do not need to consume a job. And I was about to say there is not a single place where we need more than read-only access to the job, once is has been built by the end-user, BUT there will be: in the Ent Faktory batches:
```rust
pub fn add(&mut self, mut job: Job) -> Result<Option<serde_json::Value>, Error> {
        let bid = job.custom.insert("bid".into(), self.bid.clone().into());
        self.prod.enqueue(job).map(|_| bid)
}
```
In this case we will need a mutable reference and we will be mutating the job returning the user the previous value of `bid` if there was one - so that it's not a surprise for them. So `&mut Job` would be just enough.

As for the `Cow` - this will require `Job` to be `Clone`, while it's not. My understanding was that you did not make it `Clone` initially for a reason: this would clone the `jid` and potentially can lead to duplicates, since the user may think they can simply clone a job - if they need a couple of them - and send it to the server. On the other hand, implementing `Clone` where `jid` will be re-generated - may be surprising behavior which we want to avoid in all cases. Was this assumption correct?

So, I think we can go with `& Job` and `&mut Job` in the future.